torchtune/datasets/_packed.py

```diff
@@ -119,156 +121,170 @@ def _pack(self) -> None:
-    ) -> Dict[str, torch.Tensor]:
-        """Pad a group of packed sequences to max sequence length in the group, and
-        convert integer lists to tensors. Account for attention mask and position
-        ids.
-
-        This is a sample-wise collator and should not be used with the dataloader.
-
-        Sample should look like::
-            {
-                "tokens": List[int],
-                "labels": List[int],
-                "mask": Tensor,
-                "input_pos": List[int],
-            }
-
-        Args:
-            sample (Dict[str, Any]): A dictionary containing tokens, labels, mask,
-                and position ids
-            padding_idx (int): Padding index for input ids. Defaults to 0.
-            ignore_idx (int): Padding index for labels. Defaults to -100.
-
-        Returns:
-            Collated tokens, labels, mask, and position ids.
+    def _convert_to_tensors(self, pack: PACK_TYPE) -> PACK_TYPE:
+        """Converts a pack into tensors. Pack comes in as a dict of lists and is converted to tensors.
+        The only key that does not get converted is ``seq_lens``.
         """
-        tokens = sample["tokens"]
-        labels = sample["labels"]
-        mask = sample["mask"]
-        input_pos = sample["input_pos"]
-
-        # Pad to max sequence length
-        tokens = F.pad(
-            torch.tensor(tokens), (0, self.max_seq_len - len(tokens)), value=padding_idx
-        )
-        labels = F.pad(
-            torch.tensor(labels), (0, self.max_seq_len - len(labels)), value=ignore_idx
-        )
-        # For attention mask, simply use identity matrix for the pad tokens
-        mask_pad = torch.eye(self.max_seq_len - mask.shape[0], dtype=torch.bool)
-        mask = torch.block_diag(mask, mask_pad)
-        # For position ids, continue to increment for pad tokens
-        next_pos = input_pos[-1] + 1
-        input_pos_pad = torch.arange(
-            next_pos, next_pos + self.max_seq_len - len(input_pos)
+        return {
+            "tokens": torch.tensor(pack["tokens"]),
+            "labels": torch.tensor(pack["labels"]),
+            "input_pos": torch.tensor(pack["input_pos"]),
+            "seq_lens": pack["seq_lens"],
+        }
+
+    def _pad_pack(self, pack: PACK_TYPE, padding_idx: int) -> PACK_TYPE:
+        """Pads a pack to ``self.max_seq_len``."""
+        # Pad tokens
+        padded_tokens = F.pad(
+            pack["tokens"],
+            (0, self.max_seq_len - len(pack["tokens"])),
+            value=padding_idx,
         )
-        # Do not go beyond max_seq_len - 1
-        input_pos_pad = input_pos_pad.clamp(max=self.max_seq_len - 1)
-        input_pos = torch.cat(
-            [
-                torch.tensor(input_pos),
-                input_pos_pad,
-            ]
+
+        # Pad labels
+        padded_labels = F.pad(
+            pack["labels"],
+            (0, self.max_seq_len - len(pack["labels"])),
+            value=CROSS_ENTROPY_IGNORE_IDX,
```

Comment:
It's weird that we pass padding_idx but still have `CROSS_ENTROPY_IGNORE_IDX` as a global constant. (I'm just complaining, don't worry about it here)

Reply:
Yeah I dislike this as well b/c it means that we don't actually support losses that do NOT use this ignore index OOTB. BUT, we make this assumption throughout the repo so there's no point in deviating here in this PR.